### PR TITLE
Rewrite JWT package to: target modern .NET; use up-to-date & performant dependencies; be more flexible for use.

### PR DESCRIPTION
**Pull request by Duslerke**
# What:
 - Rewrite the JWT package's core logic such that it would get decoupled from `Microsoft.AspNetCore.Http`.
 - Mark the original `TokenFactory.Create` method as in deprecation.
 - Provide the ability to parse the JWT into an arbitrary model.
 - Provide the ability to do a cheap compute check to determine the token type.
 - Ensure the token handling edge cases are all covered.
 - Extend tests coverage.
 - Drop the reliance on `Newtonsoft.Json` in favour of `System.Text.Json`.
 - Drop the reliance on `System.IdentityModel.Tokens.Jwt` in favour of  `Microsoft.IdentityModel.JsonWebTokens`.
 - Provide a package-managed Legacy Service Token model.
 - Configure additional logging visibility upon encountering token parsing failures.
 - Add JWT package documentation README.
 - Update the project to .NET 8.0 and C# language 12.
 - Configure both main and testing JWT project with nullable context.
 - Add intelisense XML documentation for JWT package interfaces, classes and models.
 - Reorganise the '.csproj` XML attributes.
 - Configure additional static code analysers and address warnings/suggestions where approapriate.

# Why:
 - The `Create` method violates the single responsibility principle in making the JWT Nuget package to both handle the JWT decoding logic as well as be aware of Http headers abstraction and perform token extraction from it. This also indirectly ties the package's use to Http APIs while it could be much more versitile. As such, the package methods were refactored to split away SRP-violating method to merely be a wrapper that's marked as becoming obsolete. Obviously, the logic of extracting the token from HTTP headers is not entirely trivial, and as such a good argument could be made that we don't want that repeated within a single API - which is why this logic should instead live in `Hackney.Core.Http`!
 - The ability to parse the JWT string to an arbitrary model _(rather than existing Token domain model)_ was added so that the Nuget package does not become exclusively tied down to the Housing systems domain.
 - The ability to determine the token type of User versus Service _(machine to machine)_ was added as our authorisation flows need to perform a different authorisation logic depending on whether it's a User's or a Machine's token. Trying to parse and decode token to user schema, and then trying to decode to machine schema on failure is a needlessly expensive in opaque _(to business logic)_ process that an now be replaced with explicit logic checks.
 - Ensured that all token decoding failures result in `null` Token being returned rather than throwing an exception _(with the minor exception to that being that dreaded HTTP headers logic that had to be kept in for backwards compatibility)_. Idea behind this is that when the token is malformed or not valid, the API's HTTP response should be a `4xx` status code signifying `user error` _(bad input)_. If we were to start throwing the errors on such tokens, we risk that the consumer of the package will not handle some of them allowing them to bubble up and result in 500 Internal Server Error _(which is not correct and is misleading)_.
 - Further extended tests coverage considering the package has barely had any worthwhile tests to begin with. A few were added during the development of Cognito token support feature #68 , however, a lot of blind spots were left to patch in the future for when the package is totally reworked.
- Switch from `Newtonsoft.Json` to `System.Text.Json` is done for performance and security reasons, plus the `System.Text.Json` is built into the .NET framework. This means external `Newtonsoft.Json` dependency can be dropped entirely, which reduces our attack surface when it comes to supply chain attacks _(also `Newtonsoft.Json` was never marked as a direct dependency even though it was getting used. It was being implicitly imported as dependency of another dependency - extremely fragile!)_.
 - Move from `System.IdentityModel.Tokens.Jwt` (`JwtSecurityTokenHandler`) to `Microsoft.IdentityModel.JsonWebTokens` (`JsonWebTokenHandler`) was done as the former legacy package is being actively phased out by Microsoft _(see [.NET 8 breaking changes documentation](https://learn.microsoft.com/en-us/aspnet/core/breaking-changes/8/securitytoken-events))_. Newer `JsonWebToken` implementation _(see [IdentityModel Wiki](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki))_, provides ~30% better performance, relies natively on `System.Text.Json`, and is optimized for modern .NET memory models _(using `Span<T>` rather substrings on heap)_ and AOT compilation _(see more [ASP.NET Core's architecture updates](https://github.com/dotnet/aspnetcore/issues/49469), and [IdentityModel 7x](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki/IdentityModel-7x))_. This also resolves the reported security vulnerability [dependabot#2](https://github.com/LBHackney-IT/lbh-core/security/dependabot/2)_(PR #65)_
 - Added a Legacy Machine Token model as it's going to be needed across multiple code-bases. It's best to have the shared tokens within a Nuget package. Though I have some doubts for whether this package is the best place for that _(though at the same time it's functionality deeply relies on those models)_.
 - Added logging to reduce the blind spots for token failures like one we had where an authorizer bug caused the token to not get passed over properly. From the API logs perspective it was not clear what the cause for unauthorised responses was.
 - Bumped the project to the .NET 8.0 target as we don't support _(nor can deploy)_ any applications that are older than that. Additionally, every subsequent .NET version has strict backwards compatibility requirements, meaning, a .NET 10 application can easily consume the .NET 8 Nuget package - this is by design to ensure older nuget version packages can still be consumed without needing to wait for them to upgrade. As such the `.netstandard2.1` compile target was unnecessary and was dropped to avoid the compile time errors when using the newer C# 12 syntax.
 - Lastly, this behing a shared Nuget package, all the documentation including the XML doc strings as well as READMEs was updated so consumers know how to use the package. Read more in Notes.
 - Reorganised the '.csproj' files to make it clearer which keys correspond to what aspect of the Nuget package so that it's easier to reason about it and maintain it.
 - Additional code analysis was enabled to ensure that this new fresh slate is as clean as it can be for the future. We don't want any legacy technical debt in the form of needlessly suppressed warnings. Also, for any warnings that remain suppressed, a documentation comment was added to explain why it was suppressed.

# Notes:
 - Thus far the `TokenFactory.Create` is being kept around as a wrapper around the core JWT package methods, therefore the public contract did not change in this regard. Where it did change, however, was in the `TokenFactory` class's constructor. To make it log token parsing failures, a logger dependency had to be added. As such, `TokenFactory` no longer has the parameter-less constructor. As far as I've seen we don't have any code that would `new TokenFactory();`, so the package should be safe to upgrade to from a typical API. Regardless of that, a major version has been bumped to signify this lack of compatibility in this regard in case any older version consumers decide to start using a parameter-less constructor.
 - The token logging has to be sanitised to prevent logs injection and forging _(`\n` and \r`)_ need be removed. Additionally, should the token would be deemed malformed or invalid _(when it is valid, but with extra unexpected characters)_, then we don't want to be accidentally leaked in the logs - as such the token value gets truncated. We're printing out the first few characters to make sure it's coming in in a bearer form and not empty or [object Object].
 - The mere addition of XML comments to the Nuget package does not make them available to the consumer of published package. By default those XML comments are only available for the maintainers of the package itself because upon code being compiled to DLLs, all the XML comments get stripped out to keep DLLs as light as possible. Based on the documentation, to get the intellisense working for the consumers of the published version of the package, the `GenerateDocumentationFile` setting needs to be enabled that takes all the XML comments from `MyFile.cs`, puts them into `MyFile.xml` and links the 2 files.
 - Lastly, the `PackageReadmeFile` was added so that the Nuget package page would have a proper and correct readme. Though at this time, I can't quite tell which of the 2 pages will get updated. It will either be the package page listing all the versions where it currently points to the wrong readme, or it will be the individual package version's page that's only displaying the package description.

# Testing:
 - Successfully tested on MMH --> RepairsAPI integration within the `development` environment logged in using Cognito token _(Repairs was failing with Unauthorized before)_ when MMH was doing a request. The test was limited to the User token flow. The Machine to Machine flow was tested on `staging`. It's only used by the Repairs Online functionality, which has `staging` and `production` environments, but not `development`. The Repairs Online was successfully able to make an API call to the Repairs API to create work order using the machine to machine token _(the 500 failures seen on dev tools come from the fact that this flow also makes SOAP API calls to DRS, which seems to be misconfigured on `staging` environment)_.

### CODE REVIEW

File: Hackney.Core/Hackney.Core.JWT/TokenFactory.cs [hunk 1]
```diff
@@ -1,76 +1,166 @@
...
-                Name = presentation.Name,
-                Nbf = presentation.Nbf,
-                Exp = presentation.Exp,
-                Iat = presentation.Iat
-            };
+            return JsonSerializer.Deserialize<T>(decodedPayload, _jsonOptions);
+        }
+        catch (Exception ex)
+        {
+            _logger.LogWarning(ex, "Unexpected, Null, or Malformed token: {InvalidToken}.", jwtStr);
```
[the remark this answers is not in the document]
Comment (Duslerke) [reply Hackney.Core/Hackney.Core.JWT/TokenFactory.cs]: Addressed and implemented.

**Review by adamtry — APPROVED** [review]
